internal/wgopenbsd/client_openbsd.go

```diff
@@ -40,9 +57,44 @@ func (c *Client) Close() error {
 
 // Devices implements wginternal.Client.
 func (c *Client) Devices() ([]*wgtypes.Device, error) {
-	// Unimplemented: no devices means this code can be built but is effectively
-	// a no-op.
-	return nil, nil
+	ifg := wgh.Ifgroupreq{
+		// Query for devices in the "wg" group.
+		Name: [16]int8{0: 'w', 1: 'g'},
+	}
+
+	// Determine how many device names we must allocate memory for.
+	if err := ioctl(c.fd, wgh.SIOCGIFGMEMB, unsafe.Pointer(&ifg)); err != nil {
+		return nil, err
+	}
+
+	// ifg.Len is in bytes; allocate enough space for the correct number
+	// of devices and then store a pointer to the address where the data should
+	// be written in the ifg.Ifgru union.
+	//
+	// TODO(mdlayher): is this actually safe? Can we guarantee that the memory
```

Comment:
This appears to work but it merits some major scrutiny. From zeebo on Gophers Slack:

> i think the usage of `ifgrs` is unsafe...
> maybe it's alright because it'll always be heap allocated due to the dynamic size of it, but afaict, the compiler could deduce that it doesn't escape and then the call to ioctl could move it during a stack realloc..